### PR TITLE
Deprecate symbol literals

### PR DESCRIPTION
**Pull request by eed3si9n**
Ref https://github.com/scala/scala-dev/issues/574

### CODE REVIEW

**Comment by eed3si9n** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/5404/console

**Comment by SethTisue** [comment]
and how about removal under `-Xsource:2.14`? see discussion at https://github.com/scala/scala-dev/issues/459

**Comment by SethTisue** [comment]
I see that you made it so that under `-Xsource:2.14`, it's an error 👍

**Comment by SethTisue** [comment]
last call for further discussion at https://github.com/scala/scala-dev/issues/459 . my sense from the discussion there is that we are likely to merge this.

**Comment by ashawley** [comment]
Seems like this should move forward given the discussion so far on dropping symbol literals.  There's a preponderance of evidence that this can work, so it seems marking symbol literals as deprecated can move forward.

I investigated the subsequent step of experimentally dropping symbol literals entirely.  I dropped them in 2.13 and backported it to 2.12.  I also used a shorthand syntax for symbols in #7395, rather than `Symbol.apply`, throughout the compiler and the standard library.  It went smoothly.

I did the exercise on 2.12 so that I could see the impact on community libraries.  At the time, the 2.13 community build was in too much flux.   With @SethTisue's assistance, I was able drop symbol literals in 24 community projects that had literal symbols in their build, see https://github.com/scala/community-builds/pull/827.  I added the new symbol syntax to these builds, and this resulted in 92 projects built successfully with a 2.12 compiler with symbol literals dropped.

There was one project of the 24 that the new symbol syntax could not be made to work, Shapeless, but there's evidence that it could be made to work with `Symbol.apply`.  Alternatively, Shapeless could undergo some significant re-architecting to shorthand symbols or could drop symbol support entirely.

There were 85 builds that didn't run in the 2.12 community build because a dependency failed to build or because Shapeless can't build.  I stopped going further only for lack of time.

**Comment by SethTisue** [comment]
@eed3si9n are you available to rebase this and fix CI for merge?

**Comment by eed3si9n** [comment]
@SethTisue Rebased against master.

**Comment by SethTisue** [comment]
@eed3si9n CI logs show that partests need updating. note that if you rebase again you'll get #7495 which will let you use `sym"foo"` if you want to

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -1311,8 +1311,13 @@ self =>
       def finish(value: Any): Tree = try newLiteral(value) finally in.nextToken()
       if (in.token == INTERPOLATIONID)
         interpolatedString(inPattern = inPattern)
-      else if (in.token == SYMBOLLIT)
+      else if (in.token == SYMBOLLIT) {
+        def msg(what: String) =
+          s"""symbol literal is $what; use Symbol("${in.strVal}") instead"""
```
Comment (SethTisue) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: you can recommend the `sym` interpolator

**Comment by ashawley** [comment]
We're only deprecating symbol literals, so there's no need to modify tests by replacing them with `Symbol` or even `sym`.†  We can look forward to doing that, or even deleting many of the tests, when symbol literals are dropped in 2.14.

I'd argue that the tests should just be updated from the deprecation.  I think this will do that:

    partest --update-check test/files/neg/checksensible.scala test/files/neg/names-defaults-neg.scala test/files/run/t10646.scala test/files/run/t4560.scala test/files/run/t4601.scala test/files/run/t6632.scala test/files/run/t6633.scala test/files/run/t6634.scala test/files/run/t7974 test/files/run/t8933 test/files/run/t8933b test/files/jvm/manifests-new.scala test/files/jvm/manifests-old.scala test/files/jvm/serialization-new.scala test/files/jvm/serialization.scala test/files/run/SymbolsTest.scala test/files/run/arrays.scala test/files/run/fors.scala test/files/run/lisp.scala

† I appreciate the `sym` evangelism.  FWIW, I did a single test of the `sym` syntax throughout the compiler in #7499 and then in the community build.  What I learned from that exercise I added to a JUnit test for `sym`.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 1 above)
Comment (eed3si9n) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: Done in 579c620

**Comment by eed3si9n** [comment]
@ashawley Thanks! That was helpful.

**Comment by ashawley** [comment]
After changing the deprecation message, you just need to update your partests:

    partest --update-check test/files/neg/symbol-literal-deprecation.scala test/files/neg/symbol-literal-removal.scala

**Comment by ashawley** [comment]
If you have it in you, you could roll back the changes to the old quoted symbol tests:

    git checkout 2.13.x -- test/files/run/SymbolsTest.scala test/files/run/arrays.scala test/files/run/fors.scala test/files/run/lisp.scala test/files/run/t10646.scala test/files/run/t8933/A_1.scala test/files/run/t8933/Test_2.scala test/files/run/t8933b/A.scala test/files/run/t8933b/Test.scala

And then update the tests:

    partest --update-check test/files/run/SymbolsTest.scala test/files/run/arrays.scala test/files/run/fors.scala test/files/run/lisp.scala test/files/run/t10646.scala test/files/run/t8933 test/files/run/t8933b

**Comment by eed3si9n** [comment]
@ashawley Are you saying you'd prefer the symbol literal kept as-is in the tests?
I think it's better to migrate the test as much as possible when we deprecate a feature to avoid future pile-up of old syntax.

**Comment by ashawley** [comment]
I understand the urge would be to migrate them for the sake of reducing technical debt.  That may be possible for certain types of deprecations, but here it would remove the code coverage of the parser for symbol literals in the 2.13 test suite for now and evermore.  I'm not sure if there are typically a lot of changes to the parser in bug-fix releases of the compiler or not, but if so I'd want to keep those symbol literal tests as-is.

**Pull request merged**